File: Algorithms/Sorting/Python/SelectionSort.py
def selection_sort(arr):
    for i in range(0, len(arr)):
        minIndex = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[minIndex]:
                minIndex = j

        temp = arr[minIndex]
        arr[minIndex] = arr[i]
        arr[i] = temp
        #arr[i], arr[minIndex] = arr[minIndex], arr[i]


# Sorting integers
numbers = [8, 20, 3, 43, 12, 125, 86, 63, 28]
selection_sort(numbers)
for sortedNumbers in numbers:
    print(sortedNumbers, end=" ")

print()
# Sorting strings
names = ["Oliver", "Nathan", "Ellie", "Jorge", "Eleni"]
selection_sort(names)
for sortedNames in names:
    print(sortedNames, end=" ")

# Note: In contrast to languages like Java and C#, Python's dynamic typing and duck typing enable this function to work
# with various data types, similar to how a generic function behaves in Java or C#.

File: Algorithms/Sorting/C#/SelectionSort.cs
using System;

public class Sorting
{
    public static void Main(string[] args)
    {
        int[] numbers = { 8, 20, 3, 43, 12, 125, 86, 63, 28 };
        string[] names = { "Oliver", "Nathan", "Ellie", "Jorge", "Eleni" };

        void SelectionSort(int[] array) //Non-Generic Selection Sort Implementation
        {
            for (int i = 0; i < array.Length; i++)
            {
                int minIndex = i;
        
                for (int j = i + 1; j < array.Length; j++)
                {
                    if (array[j] < array[minIndex]) 
                    { 
                        minIndex = j;
                    }
                }
                
                var temp = array[minIndex]; 
                array[minIndex] = array[i];
                array[i] = temp;    
            }
        }

        SelectionSort(numbers);

        foreach (int sortedNumbers in numbers)
        {
            Console.Write(sortedNumbers + " ");
        }

        Console.WriteLine();


        
        void GenericSelectionSort<T>(T[] array) where T : IComparable  //Generic Selection Sort Implementation
        {
            for (int i = 0; i < array.Length; i++)
            {
                int minIndex = i;

                for (int j = i + 1; j < array.Length; j++)
                {
                    if (array[j].CompareTo(array[minIndex]) < 0)
                    {
                        minIndex = j;
                    }
                }

                var temp = array[minIndex];
                array[minIndex] = array[i];
                array [i] = temp;   
            }
        }

        GenericSelectionSort(numbers);
        GenericSelectionSort(names);

        foreach (int sortedNumbers in numbers)
        {
            Console.Write(sortedNumbers + " ");
        }

        Console.WriteLine();

        foreach (string sortedNames in names)
        {
            Console.Write(sortedNames + " ");
        }
    }
}
